src/lib.rs

```diff
@@ -284,9 +291,142 @@ cfg_if! {
 /// In general, `getrandom` will be fast enough for interactive usage, though
 /// significantly slower than a user-space CSPRNG; for the latter consider
 /// [`rand::thread_rng`](https://docs.rs/rand/*/rand/fn.thread_rng.html).
+#[inline]
 pub fn getrandom(dest: &mut [u8]) -> Result<(), Error> {
-    if dest.is_empty() {
-        return Ok(());
+    // SAFETY: The `&mut MaybeUninit<_>` reference doesn't escape, and
+    // `getrandom_uninit` guarantees it will never de-initialize any part of
+    // `dest`.
+    getrandom_uninit(unsafe { slice_as_uninit_mut(dest) })?;
+    Ok(())
+}
+
+/// Version of the `getrandom` function which fills `dest` with random bytes
+/// returns a mutable reference to those bytes.
+///
+/// On successful completion this function is guaranteed to return a slice
+/// which points to the same memory as `dest` and has the same length.
+/// In other words, it's safe to assume that `dest` is initialized after
+/// this function has returned `Ok`.
+///
+/// No part of `dest` will ever be de-initialized at any point, regardless
+/// of what is returned.
+///
+/// # Examples
+///
+/// ```ignore
+/// # // We ignore this test since `uninit_array` is unstable.
+/// #![feature(maybe_uninit_uninit_array)]
+/// # fn main() -> Result<(), getrandom::Error> {
+/// let mut buf = core::mem::MaybeUninit::uninit_array::<1024>();
+/// let buf: &mut [u8] = getrandom::getrandom_uninit(&mut buf)?;
+/// # Ok(()) }
+/// ```
+#[inline]
+pub fn getrandom_uninit(dest: &mut [MaybeUninit<u8>]) -> Result<&mut [u8], Error> {
+    imp::getrandom_inner(dest)?;
+    // SAFETY: `dest` has been fully initialized by `imp::getrandom_inner`
+    // since it returned `Ok`.
+    Ok(unsafe { slice_assume_init_mut(dest) })
+}
+
+/// A marker indicating it is safe to use `MaybeUninit::as_bytes_mut()` to
+/// initialize a valid value of the type. Feature: `rust-1-55`.
+///
+/// This trait is implemented for byte arrays (e.g. `[u8; T]`, and levels of nested byte arrays
+///
+/// Note: `TransmuteFromArbitraryBytes` is intentionally not implemented on
+/// `u8` just because it is counterproductive to generate random values one
+/// byte at a time.
+///
+/// TODO: Should this be a sealed trait.
+#[cfg(feature = "rust-1-55")]
+pub unsafe trait TransmuteFromArbitraryBytes {}
+
+#[cfg(feature = "rust-1-55")]
+unsafe impl<const N: usize> TransmuteFromArbitraryBytes for [u8; N] {}
+
+#[cfg(feature = "rust-1-55")]
```

Comment:
I think it's worth to move `getrandom_array` and the trait into a separate module. It will allow feature-gating module and instead of duplicating `cfg`s over all items.

Also I think it's worth to use the `doc(cfg(..))` feature to make docs nicer on docs.rs.